src/API/Site/Controllers/RestAPI/AuthController.php

```diff
@@ -0,0 +1,181 @@
+	 *
+	 * @var string[]
+	 */
+	private const NEXT_PATH_MAPPING = [
+		'setup-mc'  => '/google/setup-mc',
+		'reconnect' => '/google/settings',
+	];
+
+	/**
+	 * AuthController constructor.
+	 *
+	 * @param RESTServer $server
+	 */
+	public function __construct( RESTServer $server ) {
+		parent::__construct( $server );
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 */
+	public function register_routes() {
+		$this->register_route(
+			'rest-api/authorize',
+			[
+				[
+					'methods'             => TransportMethods::READABLE,
+					'callback'            => $this->get_authorize_callback(),
+					'permission_callback' => $this->get_permission_callback(),
+					'args'                => $this->get_auth_params(),
+				],
+			]
+		);
+	}
+
+	/**
+	 * Get the callback function for the authorization request.
+	 *
+	 * @return callable
+	 */
+	protected function get_authorize_callback(): callable {
+		return function ( Request $request ) {
+			/*
+			The full auth URL example:
+
+			https://public-api.wordpress.com/oauth2/authorize?
+			client_id=CLIENT_ID&
+			redirect_uri=PARTNER_REDIRECT_URL&
+			response_type=code&
+			blog=BLOD_ID&
+			scope=wc-partner-access&
+			state=BASE64_ENCODED_STRING
+
+			Content of state is an URL query string where the value of its parameter "redirect_url" is being URL encoded.
+			E.g.
+			nonce=nonce-123&redirect_url=https%3A%2F%2Fmerchant-site.example.com%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin%26path%3D%2Fgoogle%2Fsetup-mc
+
+			where its URL decoded version is:
+			nonce=nonce-123&redirect_url=https://merchant-site.example.com/wp-admin/admin.php?page=wc-admin&path=/google/setup-mc
+
+			Ref: https://developer.wordpress.com/docs/oauth2/
+			*/
+
+			try {
+				$auth_url = 'https://public-api.wordpress.com/oauth2/authorize';
+
+				// TODO: Call an API by Google with merchant information and get the below data.
+				// We'd probably need use WCS to communicate with the new API.
+				$client_id    = '91299';
+				$redirect_uri = 'https://woo.com';
+				$nonce        = 'nonce-123';
+
+				$response_type = 'code';
+				$blog_id       = Jetpack_Options::get_option( 'id' );
+				$scope         = 'wc-partner-access';
+
+				$next                  = $request->get_param( 'next_page_name' );
+				$path                  = self::NEXT_PATH_MAPPING[ $next ];
+				$merchant_redirect_url = urlencode_deep( admin_url( "admin.php?page=wc-admin&path={$path}" ) );
+
+				$state = base64_encode(
+					build_query(
+						[
+							'nonce'        => $nonce,
+							'redirect_url' => $merchant_redirect_url,
+						]
+					)
+				);
+
+				$auth_url = esc_url_raw(
+					add_query_arg(
+						[
+							'blog'          => $blog_id,
+							'client_id'     => $client_id,
+							'redirect_uri'  => $redirect_uri,
+							'response_type' => $response_type,
+							'scope'         => $scope,
+							'state'         => $state,
+						],
+						$auth_url
+					)
+				);
```

Comment:
Thanks for the suggestion @mikkamp.

> Not sure if this PR is fully ready for a review yet, as a lot of it seems preliminary

Since this PR is going to be merged into a feature branch rather than `develop` so I was thinking using the hardcoded data first for testing and revisit it when we get the information about the new Google API.

> Can we move that to a service class which can just have the function of building an auth URL

Completely agreed, why didn't I do it... 😓 Working on it.

> We could consider adding it in src/API/WP. The NotificationService seems to have been added to src/Google but maybe we'd want to consider having them grouped together, since they both do similar things using the blog_id and handling the URL to public-api.wordpress.com.

Great idea, thanks.